src/MatrixFields/field_name_dict.jl

```diff
@@ -24,52 +24,51 @@ matrix can be computed by calling `one(dict)`.
 - Addition and subtraction
-- Multiplication, where the first argument must be a `FieldMatrix` (or
-  `FieldMatrixBroadcasted`)
-- Inversion, where the argument must be a diagonal `FieldMatrix` (or
-  `FieldMatrixBroadcasted`), i.e., one in which every entry is either a
-  `ColumnwiseBandMatrixField` of `DiagonalMatrixRow`s or a multiple of
-  `LinearAlgebra.I`
+- Multiplication, where the first argument must be a `FieldMatrix`
+- Inversion, where the argument must be a diagonal `FieldMatrix`, i.e., one in
+  which every entry is either a `ColumnwiseBandMatrixField` of
+  `DiagonalMatrixRow`s or a multiple of `LinearAlgebra.I`
 """
-struct FieldNameDict{T1, T2, K <: FieldNameSet{T1}, E <: NTuple{<:Any, T2}} <:
-       AbstractDict{T1, T2}
+struct FieldNameDict{
+    T <: Union{FieldName, FieldNamePair},
+    K <: FieldNameSet{T},
+    E,
+} <: AbstractDict{T, Any}
     keys::K
     entries::E
 
     # This needs to be an inner constructor to prevent Julia from automatically
     # generating a constructor that fails Aqua.detect_unbound_args_recursively.
-    FieldNameDict{T1, T2}(
-        keys::FieldNameSet{T1, <:NTuple{N, T1}},
-        entries::NTuple{N, T2},
-    ) where {T1, T2, N} =
-        new{T1, T2, typeof(keys), typeof(entries)}(keys, entries)
+    function FieldNameDict(keys::FieldNameSet{T}, entries) where {T}
+        length(keys) == length(entries) || error(
+            "FieldNameDict cannot have different numbers of keys and entries",
+        )
+        unrolled_foreach(entries) do entry
+            check_entry(T, entry) ||
+                error("Invalid $(FieldNameDict{T}) entry: $entry")
+        end
+        return new{T, typeof(keys), typeof(entries)}(keys, entries)
+    end
 end
-function FieldNameDict{T1, T2}(
-    key_entry_pairs::Pair{<:T1, <:T2}...;
-    name_tree = nothing,
-) where {T1, T2}
+function FieldNameDict{T}(key_entry_pairs::Pair{<:T}...) where {T}
     keys = unrolled_map(pair -> pair[1], key_entry_pairs)
     entries = unrolled_map(pair -> pair[2], key_entry_pairs)
-    return FieldNameDict{T1, T2}(FieldNameSet{T1}(keys, name_tree), entries)
+    return FieldNameDict(FieldNameSet{T}(keys), entries)
 end
-FieldNameDict{T1}(args...) where {T1} = FieldNameDict{T1, Any}(args...)
-
-const FieldVectorView = FieldNameDict{FieldName, Fields.Field}
-const FieldVectorViewBroadcasted =
-    FieldNameDict{FieldName, Union{Fields.Field, Base.AbstractBroadcasted}}
-const FieldMatrix = FieldNameDict{
-    FieldNamePair,
-    Union{UniformScaling, ColumnwiseBandMatrixField},
-}
-const FieldMatrixBroadcasted = FieldNameDict{
-    FieldNamePair,
-    Union{UniformScaling, ColumnwiseBandMatrixField, Base.AbstractBroadcasted},
-}
 
-dict_type(::FieldNameDict{T1, T2}) where {T1, T2} = FieldNameDict{T1, T2}
+const FieldVectorView = FieldNameDict{FieldName}
+const FieldMatrix = FieldNameDict{FieldNamePair}
+
+check_entry(_, _) = false
+check_entry(::Type{FieldName}, ::Fields.Field) = true
```

Comment:
```suggestion
check_entry(::Type{<:FieldName}, ::Fields.Field) = true
```
Since `FieldName` has a parameter type, does this specialization need to subtype `FieldName`?

Reply:
I'm actually just using `FieldName` and `FieldNamePair` as flags that switch between a vector-like dict and a matrix-like dict. I'm not using concrete `FieldName` or `FieldNamePair` types, like `FieldName{(:c, :ρ)}`, as parameters. I could equivalently replace `FieldName` and `FieldNamePair` throughout this file with symbols like `:M` and `:V`, and pass them around by wrapping them in `Val`s. Not sure if that would make things any clearer, but it would probably be the same performance-wise.